File: widgets/network.cpp
#include "include/widgets/network.hpp"
#include "include/utils/icons.hpp"

namespace widgets {

Network::Network(void *ctx)
    : Gtk::Button(), utils::Subscriber(ctx), popover(ctx) {
  label.set_label("--");

  image.set(utils::Icons::wifi_icon());

  download_speed_label.set_label("??");
  download_speed_label.set_css_classes({"network-speed-label"});
  download_speed_icon.set(utils::Icons::download_speed_icon());

  upload_speed_label.set_label("??");
  upload_speed_label.set_css_classes({"network-speed-label"});
  upload_speed_icon.set(utils::Icons::upload_speed_icon());

  Gtk::Box network_wrapper(Gtk::Orientation::HORIZONTAL, 0);
  network_wrapper.append(label);
  network_wrapper.append(image);

  Gtk::Separator sep(Gtk::Orientation::VERTICAL);
  network_wrapper.append(sep);

  network_wrapper.append(download_speed_label);
  network_wrapper.append(download_speed_icon);
  network_wrapper.append(upload_speed_label);
  network_wrapper.append(upload_speed_icon);

  set_css_classes({"widget", "network", "padded", "clickable"});
  set_name("Network");
  set_cursor("pointer");
  set_child(network_wrapper);

  popover.set_parent(*this);

  signal_clicked().connect([this]() { this->popover.popup(); });
}

void Network::on_io_event(io::Event::WifiStatus_Body data) {
  if (data.wifi_status.tag == io::COption<io::WifiStatus>::Tag::None) {
    image.hide();
    label.set_label("Not connected");
  } else {
    image.show();
    char buffer[100];
    sprintf(buffer, "%s (%d)%% ", data.wifi_status.some._0.ssid,
            data.wifi_status.some._0.strength);
    label.set_label(buffer);
  }
}
void Network::on_io_event(io::Event::NetworkSpeed_Body data) {
  download_speed_label.set_label(data.download_speed);
  upload_speed_label.set_label(data.upload_speed);
}
void Network::on_io_event(io::Event::NetworkList_Body data) {
  popover.update(data.list);
}

} // namespace widgets

File: ui/top_bar/network.py
from gi.repository import Gdk, Gio, GLib, Gtk
from utils.commands import Commands
from utils.context import ctx


class Network(Gtk.Button):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)
        ctx.pub_sub.subscribe(self)

        self.set_css_classes(["widget", "network", "padded", "clickable"])
        self.set_name("Network")
        self.set_cursor(Gdk.Cursor.new_from_name("pointer"))

        self.label = Gtk.Label.new("-- ")

        self.image = Gtk.Image.new_from_gicon(ctx.icons.wifi)

        self.download_speed_label = Gtk.Label.new("??")
        self.download_speed_label.set_css_classes(["network-speed-label"])
        self.download_speed_icon = Gtk.Image.new_from_gicon(ctx.icons.download)

        self.upload_speed_label = Gtk.Label.new("??")
        self.upload_speed_label.set_css_classes(["network-speed-label"])
        self.upload_speed_icon = Gtk.Image.new_from_gicon(ctx.icons.upload)

        network_wrapper = Gtk.Box(orientation=Gtk.Orientation.HORIZONTAL, spacing=0)
        self.set_child(network_wrapper)
        network_wrapper.append(self.label)
        network_wrapper.append(self.image)

        sep = Gtk.Separator.new(Gtk.Orientation.VERTICAL)
        network_wrapper.append(sep)

        network_wrapper.append(self.download_speed_label)
        network_wrapper.append(self.download_speed_icon)
        network_wrapper.append(self.upload_speed_label)
        network_wrapper.append(self.upload_speed_icon)

        self.popover = NetworkPopover()
        self.popover.set_parent(self)

        self.connect("clicked", self.on_click)

    def on_wifi_status(self, event):
        if event.wifi_status is None:
            self.image.hide()
            self.label.set_label("Not connected")
            return
        ssid = event.wifi_status.ssid
        strength = event.wifi_status.strength
        self.label.set_label(f"{ssid} ({strength})% ")

    def on_network_speed(self, event):
        self.download_speed_label.set_label(event.download_speed)
        self.upload_speed_label.set_label(event.upload_speed)

    def on_network_list(self, event):
        self.popover.update(event.list)

    def on_click(self, _):
        self.popover.popup()


class NetworkPopover(Gtk.PopoverMenu):
    def __init__(self, *args, **kwargs):
        super().__init__(*args, **kwargs)

        self.model = Gio.Menu.new()
        self.add_settings_row()
        self.add_ping_row()

        self.set_menu_model(self.model)
        self.set_has_arrow(False)

        action_group = Gio.SimpleActionGroup.new()

        action = Gio.SimpleAction.new("settings")
        action.connect("activate", self.on_settings_row_clicked)
        action_group.add_action(action)

        action = Gio.SimpleAction.new("ping")
        action.connect("activate", self.on_ping_row_clicked)
        action_group.add_action(action)

        action = Gio.SimpleAction.new("clicked", GLib.VariantType.new("s"))
        action.connect("activate", self.on_network_row_clicked)
        action_group.add_action(action)

        self.insert_action_group("network", action_group)

    def update(self, networks):
        self.model.remove_all()
        for network in networks:
            iface = network.iface
            address = network.address
            item = Gio.MenuItem.new(f"{iface}: {address}", None)
            item.set_action_and_target_value(
                "network.clicked", GLib.Variant("s", address)
            )
            self.model.append_item(item)
        self.add_settings_row()
        self.add_ping_row()

    def add_settings_row(self):
        item = Gio.MenuItem.new("Settings (nmtui)", "network.settings")
        self.model.append_item(item)

    def add_ping_row(self):
        item = Gio.MenuItem.new("Ping", "network.ping")
        self.model.append_item(item)

    def on_settings_row_clicked(self, action, parameter):
        Commands.spawn_network_editor()

    def on_ping_row_clicked(self, action, parameter):
        ctx.windows.ping.toggle()

    def on_network_row_clicked(self, action, parameter):
        ip = parameter.get_string()

        display = Gdk.Display.get_default()
        clipboard = display.get_clipboard()
        clipboard.set(ip)

        notification = Gio.Notification.new(f"Copied {ip}")
        ctx.app.send_notification(None, notification)
